# Exploratory Data Analysis (EDA) 1.0 - Metadata

In this section, we conduct an exploratory data analysis of the CBIS-DDSM Case and DICOM datasets prepared in prior sections. The purpose of this EDA is three-fold:

1. Discover the relationships among the features
2. Explore the nature of the relationships between the features and the diagnostic target.
3. Identify features that have the greatest influence classification accuracy.

## The Dataset

In this section, we will be analyzing the Case Dataset: calcification and mass datasets containing patient, abnormality, BI-RADS assessment, image image_view, breast density, subtlety, and pathology (diagnosis) information. The dataset dictionary is outlined in {numref}`eda1_case_dataset`.

```{table} Case Dataset Dictionary
:name: eda1_case_dataset

| #  | Variable             | Type        | Description                                                                                                                  |
|----|----------------------|-------------|------------------------------------------------------------------------------------------------------------------------------|
| 1  | patient_id           | Nominal     | Unique identifier for each patient.                                                                                          |
| 2  | breast_density       | Discrete    | BI-RADS overall assessment of the volume of attenuating tissues in the breast.                                             |
| 3  | left_or_right_breast | Nominal     | Which breast was imaged.                                                                                                     |
| 4  | image_view           | Dichotomous | Either cranialcaudal or mediolateral oblique image_view.                                                                            |
| 5  | abnormality_id       | Discrete    | Number of abnormalities for the patient.                                                                                       |
| 6  | abnormality_type     | Dichotomous | BI-RADS category of the abnormality.                                                                                         |
| 7  | calc_type            | Nominal     | Characterization of the type of calcification (where applicable)                                                             |
| 8  | calc_distribution    | Nominal     | The arrangement of the calcifications inside the breast and, relative to the probability of malignancy. (where applicable) |
| 9  | mass_shape           | Nominal     | Shape of the mass                                                                                                            |
| 10 | mass_margins         | Nominal     | Feature that separates the mass from the adjacent breast parenchyma.                                                         |
| 11 | assessment           | Discrete    | Overall BI-RADS assessment of the mammography                                                                                |
| 12 | pathology            | Nominal     | Determination of the malignancy of the case.                                                                                 |
| 13 | subtlety             | Discrete    | Degree of diagnostic difficulty                                                                                              |
| 14 | fileset              | Nominal     | Indicates training or test set.                                                                                              |
| 15 | case_id              | Nominal     | Unique identifier for the case.                                                                                              |
| 16 | cancer               | Dichotomous | Indicates whether the cancer is diagnosed.                                                                                   |
```

## Guiding Questions

```{tip}
“Far better an approximate answer to the right question, which is often vague, than an exact answer to the wrong question, which can always be made precise.” — John Tukey
```

Here, we'll put forward a set of questions to motivate and guide the discovery process.

1. What are the relationships between calcification and mass morphological features and malignancy?
2. To what degree does breast density relate to abnormality types and malignancy?
3. Are certain abnormalities more or less subtle?
4. What are the relative features of importance concerning screening?

## Exploratory Data Analysis Plan

The EDA will be conducted in three primary stages:

1. **Univariate**: Examination of the variables independently
2. **Bivariate**: Evaluate the relations among the features and between the features and the target.
3. **Multivariate**: Discover feature importance w.r.t. screening and diagnosis.

In [1]:
import sys
import os
if 'jbook' in os.getcwd():
    os.chdir(os.path.abspath(os.path.join("../../..")))
import warnings
warnings.filterwarnings("ignore")

In [2]:
import pandas as pd
from scipy import stats
import pickle
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from myst_nb import glue
import seaborn.objects as so
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

from bcd.explore.meta.case import CaseExplorer
from bcd.explore.meta.multivariate.pipeline import PipelineBuilder
from bcd.explore.meta.multivariate.selection import ModelSelector

pd.options.display.max_rows = 999
pd.options.display.max_columns = 999
sns.set_style('whitegrid')
sns.set_palette("Blues_r")

In [3]:
CASE_FP = "data/meta/3_cooked/cases.csv"

In [4]:
cases = CaseExplorer(filepath=CASE_FP)
calc = CaseExplorer(df=cases.get_calc_data())
mass = CaseExplorer(df=cases.get_mass_data())

## Univariate Analysis

Let's get an overall sense of the data.

In [5]:
summary = cases.summary()
glue("eda1_summary", summary)

Summary
Patients                            1566
Cases                               3566
Calcification Cases                 1872
Calcification Cases - Benign        1199
Calcification Cases - Malignant      673
Mass Cases                          1694
Mass Cases - Benign                  910
Mass Cases - Malignant               784

```{glue:figure} eda1_summary
---
align: center
name: eda1_summary_df
---
CBIS-DDSM Dataset Summary
```

In [6]:
st = summary.T
pct_calc = round(st['Calcification Cases'] / st['Cases'] * 100,2).values[0]
pct_mass = round(st['Mass Cases'] / st['Cases'] * 100,2).values[0]
pct_calc_mal = round(st['Calcification Cases - Malignant'] / st['Calcification Cases'] * 100,2).values[0]
pct_calc_bn = round(st['Calcification Cases - Benign'] / st['Calcification Cases'] * 100,2).values[0]
pct_mass_mal = round(st['Mass Cases - Malignant'] / st['Mass Cases'] * 100,2).values[0]
pct_mass_bn = round(st['Mass Cases - Benign'] / st['Mass Cases'] * 100,2).values[0]
cases_per_patient = round(st['Cases'] / st['Patients'],2).values[0]

glue("pct_calc", pct_calc)
glue("pct_mass", pct_mass)
glue("pct_calc_mal", pct_calc_mal)
glue("pct_calc_bn", pct_calc_bn)
glue("pct_mass_mal", pct_mass_mal)
glue("pct_mass_bn", pct_mass_bn)
glue("cases_per_patient", cases_per_patient)

52.5

47.5

35.95

64.05

46.28

53.72

2.28

From {numref}`eda1_summary_df`, several observations can be made:

1. We have 3566 cases, {glue:}`pct_calc`% are calcification cases and {glue:}`pct_mass`% are mass cases.
2. Of the calcification cases, {glue:}`pct_calc_bn`% are benign and {glue:}`pct_calc_mal`% are malignant.
3. Of the mass cases, {glue:}`pct_mass_bn`% are benign and {glue:}`pct_mass_mal`% are malignant.
4. On average, we have approximately {glue:}`cases_per_patient` cases per patient.

Case, as defined in {cite}`leeCuratedMammographyData2017`, indicates a particular abnormality, seen on the craniocaudal (CC) and/or mediolateral oblique (MLO) views.

Let's take a look at the calcification and mass data.

In [7]:
cases.get_calc_data().sample(5)
cases.get_mass_data().sample(5)

case_id patient_id  breast_density  \
48     Calcification-Train_P_00031_LEFT_CC_1    P_00031               4   
1553   Calcification-Test_P_00041_LEFT_MLO_2    P_00041               1   
46    Calcification-Train_P_00030_RIGHT_CC_1    P_00030               2   
1804  Calcification-Test_P_01635_RIGHT_MLO_2    P_01635               2   
318    Calcification-Train_P_00455_LEFT_CC_1    P_00455               1   

     left_or_right_breast image_view  abnormality_id              calc_type  \
48                   LEFT         CC               1            PLEOMORPHIC   
1553                 LEFT        MLO               2        LUCENT_CENTERED   
46                  RIGHT         CC               1               VASCULAR   
1804                RIGHT        MLO               2  AMORPHOUS-PLEOMORPHIC   
318                  LEFT         CC               1        LUCENT_CENTERED   

     calc_distribution  assessment                pathology  subtlety fileset  \
48           CLUSTERED           4                MALIGNANT         1   train   
1553         SEGMENTAL           2  BENIGN_WITHOUT_CALLBACK         5    test   
46           SEGMENTAL           2  BENIGN_WITHOUT_CALLBACK         4   train   
1804         CLUSTERED           4                MALIGNANT         3    test   
318          SEGMENTAL           2  BENIGN_WITHOUT_CALLBACK         3   train   

      cancer  
48      True  
1553   False  
46     False  
1804    True  
318    False

case_id patient_id  breast_density  \
2060    Mass-Train_P_00279_LEFT_CC_1    P_00279               3   
1952    Mass-Train_P_00106_LEFT_CC_3    P_00106               1   
2347   Mass-Train_P_00695_RIGHT_CC_1    P_00695               2   
2650   Mass-Train_P_01134_RIGHT_CC_1    P_01134               2   
2758  Mass-Train_P_01268_RIGHT_MLO_1    P_01268               3   

     left_or_right_breast image_view  abnormality_id  \
2060                 LEFT         CC               1   
1952                 LEFT         CC               3   
2347                RIGHT         CC               1   
2650                RIGHT         CC               1   
2758                RIGHT        MLO               1   

                    mass_shape   mass_margins  assessment  \
2060                     ROUND       OBSCURED           0   
1952                LYMPH_NODE  CIRCUMSCRIBED           2   
2347  ASYMMETRIC_BREAST_TISSUE    ILL_DEFINED           5   
2650                 IRREGULAR     SPICULATED           5   
2758                 LOBULATED       OBSCURED           0   

                    pathology  subtlety fileset  cancer  
2060                   BENIGN         4   train   False  
1952  BENIGN_WITHOUT_CALLBACK         4   train   False  
2347                MALIGNANT         5   train    True  
2650                MALIGNANT         5   train    True  
2758                   BENIGN         5   train   False

Our univariate analysis will cover:

- Breast Density
- Left or Right Breast
- Image View
- Abnormality Id
- Abnormality Type
- Subtlety
- BI-RADS Assessment
- Calcification Type
- Calcification Distribution
- Mass Shape
- Mass Margins
- Pathology
- Cancer (Target)